lib/OpenQA/WebAPI/Controller/Test.pm

```diff
@@ -569,18 +569,48 @@ sub prepare_job_results {
       = $schema->resultset('TestSuites')->search({name => \%desc_args}, {columns => [qw(name description)]});
     my %descriptions = map { $_->name => $_->description } @descriptions;
 
-    foreach my $job (@$jobs) {
-        next if $states         && !$states->{$job->state};
-        next if $results        && !$results->{$job->result};
-        next if $archs          && !$archs->{$job->ARCH};
-        next if $machines       && !$machines->{$job->MACHINE};
-        next if $failed_modules && $job->result ne OpenQA::Jobs::Constants::FAILED;
-
-        my $result = $job->overview_result($comment_data, $aggregated, $failed_modules, $self->param('todo')) or next;
+    my @wanted_jobs = grep {
+              (not $states         or $states->{$_->state})
+          and (not $results        or $results->{$_->result})
+          and (not $archs          or $archs->{$_->ARCH})
+          and (not $machines       or $machines->{$_->MACHINE})
+          and (not $failed_modules or $_->result eq OpenQA::Jobs::Constants::FAILED)
+    } @$jobs;
+    my @jobids                = map { $_->id } @wanted_jobs;
+    my $failed_modules_by_job = $schema->resultset('JobModules')->search(
+        {job_id => {-in => [@jobids]}, result   => 'failed'},
+        {select => [qw(name job_id)],  order_by => 't_updated'},
+    );
+    my %failed_modules_by_job;
+    for my $item ($failed_modules_by_job->all) {
+        push @{$failed_modules_by_job{$item->job_id}}, $item->name;
+    }
```

Comment:
can that be a `map`? Like

```suggestion
    my %failed_modules_by_job = map { $_->job_id => $item->name } $failed_modules_by_job->all;
```

Reply:
No, your suggestion is not going to work. The value is not a module name, but an array of module names, hence the `push`.
It could still be changed into a map, though.
```perl
map { push @{$failed_modules_by_job{$_->job_id}}, $_->name } $failed_modules_by_job->all;
```
or even
```perl
push @{$failed_modules_by_job{$_->job_id}}, $_->name for $failed_modules_by_job->all;
```
You know that I'm not convinced that short is always == readable, but if the majority prefers that I'll change it. :)

Reply:
It's not only about "saving some lines" but I think loop iterations don't reflect properly what we want to achieve: We want to initialize `%failed_modules_by_job` with data, not "go over every element and do stuff"

Reply:
ok, but your suggestion won't work.
Should I change the code to one of my suggestions, and if yes, which? :)

Reply:
If this could be a `map` I'd also be interested in how? Note that there's an array per job ID, not per element the loop iterates over.

Reply:
Use the suffix-for

Reply:
done